src/Pact/Eval.hs

```diff
@@ -121,14 +120,42 @@ enforceKeySet i ksn KeySet{..} = do
 {-# INLINE enforceKeySet #-}
 
 
--- Hoist Name back to ref
+-- | Hoist Name back to ref
 liftTerm :: Term Name -> Term Ref
 liftTerm a = TVar (Direct a) def
 
-enscopeApply :: App (Term Name) -> Eval e (Term Name)
-enscopeApply a = do
-  a' <- enscope $ TApp a (_appInfo a)
-  reduce a'
+-- | Eval a function by name with supplied args, and guard against recursive execution.
+evalByName :: Name -> [Term Name] -> Info -> Eval e (Term Name)
+evalByName n as i = do
+
+  -- Build and resolve TApp
+
+  app <- enscope (TApp (App (TVar n def) as i) i)
+
+  -- lens into user function if any to test for loop
+
+  case preview (tApp . appFun . tVar . _Ref . tDef) app of
+    Nothing -> return ()
+    Just Def{..} -> do
+
+      -- lens into call stack to find first app with matching name/module, blowup if found
+
+      let sameName :: DefName -> ModuleName -> FunApp -> Maybe ()
+          sameName dn mn FunApp{..}
+            | (DefName _faName) == dn && Just mn == _faModule = Just ()
+            | otherwise = Nothing
+
+      found <- uses evalCallStack $
+               preview (traverse . sfApp . _Just . _1 . to (sameName _dDefName _dModule) . _Just)
+
+      case found of
+        Just () -> evalError i $ "evalByName: loop detected: " <> pretty n
+        _ -> return ()
```

Comment:
we could just use `Bool` instead of `Maybe ()`. Then that extra `_Just` goes away, and we can finish with a `when`

Reply:
Tried that, it doesn't function properly -- you need `_True` or something